File: tools/convert_model_to_arrays.cpp
// ============================================================
// convert_model_to_arrays.cpp
// ============================================================
// 把 Live2D 模型目录下的所有文件（.moc3 / .json / .png / .wav /
// .motion3.json 等）编译成 C 数组，生成一个 .h 头文件，供 Android
// NDK 项目从内存直接加载，无需把文件放进 assets 目录。
//
// 这是 convert_model_to_arrays.py 的 C++ 交互式版本。
//
// 编译 (Linux/macOS):
//     g++ -std=c++17 -O2 -o convert_model_to_arrays convert_model_to_arrays.cpp
//
// 编译 (Windows, 需要 VS 或 MinGW):
//     g++ -std=c++17 -O2 -o convert_model_to_arrays.exe convert_model_to_arrays.cpp
//
// 运行:
//     ./convert_model_to_arrays
//     程序会交互式询问: 模型根目录 / 模型名 / 输出头文件路径
// ============================================================

#include <algorithm>
#include <cctype>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#if defined(_WIN32)
#include <direct.h>
#include <io.h>
#else
#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#endif

// ------------------------------------------------------------------
// 工具函数
// ------------------------------------------------------------------

// 把任意路径/文件名转成合法的 C 标识符片段。
// 例如: "Haru/Haru.2048/texture_00.png" -> "Haru_Haru_2048_texture_00_png"
static std::string SanitizeIdentifier(const std::string& name)
{
    std::string s;
    bool lastUnderscore = false;
    for (char c : name)
    {
        if (std::isalnum(static_cast<unsigned char>(c)))
        {
            s += c;
            lastUnderscore = false;
        }
        else
        {
            if (!lastUnderscore && !s.empty())
            {
                s += '_';
                lastUnderscore = true;
            }
        }
    }
    // 去掉末尾下划线
    while (!s.empty() && s.back() == '_')
        s.pop_back();
    return s;
}

// 把字节串转成 16 字节一行的 C 数组初始化文本。
static std::string BytesToCArray(const std::vector<unsigned char>& data)
{
    std::ostringstream oss;
    const size_t perLine = 16;
    for (size_t i = 0; i < data.size(); i += perLine)
    {
        oss << "    ";
        size_t end = std::min(i + perLine, data.size());
        for (size_t j = i; j < end; ++j)
        {
            char buf[8];
            std::snprintf(buf, sizeof(buf), "0x%02x", data[j]);
            oss << buf;
            if (j + 1 < end)  // 行内元素之间用 ", " 分隔
                oss << ", ";
        }
        oss << ",\n";  // 每行末尾一个逗号 (trailing comma, C 合法)
    }
    return oss.str();
}

// 读取整个文件到字节向量。失败返回 false。
static bool ReadFileBytes(const std::string& path, std::vector<unsigned char>& out)
{
    std::ifstream f(path, std::ios::binary);
    if (!f)
        return false;
    f.seekg(0, std::ios::end);
    std::streamoff len = f.tellg();
    if (len < 0)
        return false;
    f.seekg(0, std::ios::beg);
    out.resize(static_cast<size_t>(len));
    if (len > 0)
        f.read(reinterpret_cast<char*>(out.data()), len);
    return f.good() || f.eof();
}

// ------------------------------------------------------------------
// 目录遍历
// ------------------------------------------------------------------

struct FileEntry
{
    std::string relPath;  // 相对模型根目录的路径，使用 '/' 分隔
    std::string absPath;  // 绝对路径
};

// 递归收集目录下所有文件。excludeAbs: 需要排除的绝对路径（输出头文件本身）。
static void CollectFilesRecursive(const std::string& dir,
                                  const std::string& baseDir,
                                  const std::string& excludeAbs,
                                  std::vector<FileEntry>& out)
{
#if defined(_WIN32)
    // Windows 使用 _findfirst / _findnext
    std::string pattern = dir + "/*";
    struct _finddata_t fd;
    intptr_t handle = _findfirst(pattern.c_str(), &fd);
    if (handle == -1)
        return;
    do
    {
        std::string name = fd.name;
        if (name == "." || name == "..")
            continue;
        std::string full = dir + "/" + name;
        if (fd.attrib & _A_SUBDIR)
        {
            CollectFilesRecursive(full, baseDir, excludeAbs, out);
        }
        else
        {
            if (!excludeAbs.empty() && full == excludeAbs)
                continue;
            std::string rel = full.substr(baseDir.size() + 1);
            std::replace(rel.begin(), rel.end(), '\\', '/');
            out.push_back({rel, full});
        }
    } while (_findnext(handle, &fd) == 0);
    _findclose(handle);
#else
    // POSIX 使用 opendir / readdir
    DIR* d = opendir(dir.c_str());
    if (!d)
        return;
    struct dirent* ent;
    while ((ent = readdir(d)) != nullptr)
    {
        std::string name = ent->d_name;
        if (name == "." || name == "..")
            continue;
        std::string full = dir + "/" + name;
        struct stat st;
        if (stat(full.c_str(), &st) != 0)
            continue;
        if (S_ISDIR(st.st_mode))
        {
            CollectFilesRecursive(full, baseDir, excludeAbs, out);
        }
        else if (S_ISREG(st.st_mode))
        {
            if (!excludeAbs.empty() && full == excludeAbs)
                continue;
            std::string rel = full.substr(baseDir.size() + 1);
            out.push_back({rel, full});
        }
    }
    closedir(d);
#endif
}

// ------------------------------------------------------------------
// 头文件生成
// ------------------------------------------------------------------

static std::string GenerateHeader(const std::string& modelName,
                                  const std::vector<FileEntry>& files)
{
    std::string modelId = SanitizeIdentifier(modelName);
    std::string modelIdLower = modelId;
    std::transform(modelIdLower.begin(), modelIdLower.end(), modelIdLower.begin(),
                   [](unsigned char c) { return static_cast<char>(std::tolower(c)); });

    std::string findFunc = "Find" + modelId + "EmbeddedFile";
    std::string tableName = "g_" + modelIdLower + "_embeddedFiles";
    std::string countName = "g_" + modelIdLower + "_embeddedFileCount";

    std::ostringstream out;
    out << "#pragma once\n\n";
    out << "/**\n";
    out << " * Auto-generated Live2D model data header.\n";
    out << " *\n";
    out << " * All model files for '" << modelName << "' are embedded as C arrays\n";
    out << " * for in-memory loading. No assets directory required at runtime.\n";
    out << " *\n";
    out << " * Do not edit manually -- regenerate with convert_model_to_arrays.cpp.\n";
    out << " */\n\n";
    out << "#include <cstring>\n";
    out << "#include <cstddef>\n\n";

    std::vector<std::string> entryLines;
    entryLines.reserve(files.size());

    for (const auto& fe : files)
    {
        std::vector<unsigned char> data;
        if (!ReadFileBytes(fe.absPath, data))
        {
            std::cerr << "  [WARN] 无法读取文件，跳过: " << fe.relPath << "\n";
            continue;
        }

        // 变量名: g_<model>_<sanitized rel path>，例如 g_haru_Haru_moc3
        // 若相对路径以 "<模型名>/" 开头，则去掉该前缀，避免变量名重复模型名。
        std::string relForVar = fe.relPath;
        std::string prefix = modelName + "/";
        if (relForVar.compare(0, prefix.size(), prefix) == 0)
            relForVar = relForVar.substr(prefix.size());

        std::string varName = "g_" + modelIdLower + "_" + SanitizeIdentifier(relForVar);
        std::string sizeName = varName + "_size";

        out << "// Source: " << fe.relPath << "\n";
        out << "static const unsigned char " << varName << "[] = {\n";
        out << BytesToCArray(data);
        out << "};\n";
        out << "static const unsigned int " << sizeName << " = " << data.size() << ";\n\n";

        entryLines.push_back("    { \"" + fe.relPath + "\", " + varName + ", " + sizeName + " },");
    }

    // 查找表
    out << "// ============================================================\n";
    out << "// Lookup Table\n";
    out << "// ============================================================\n\n";
    out << "struct Live2DEmbeddedFile {\n";
    out << "    const char* path;           // Lookup path (e.g., \"" << (files.empty() ? std::string("Haru/Haru.model3.json") : files[0].relPath) << "\")\n";
    out << "    const unsigned char* data;  // Pointer to embedded data\n";
    out << "    unsigned int size;           // Size in bytes\n";
    out << "};\n\n";
    out << "static const Live2DEmbeddedFile " << tableName << "[] = {\n";
    for (const auto& line : entryLines)
        out << line << "\n";
    out << "};\n\n";
    out << "static const int " << countName << " = " << entryLines.size() << ";\n\n";

    // 查找函数
    out << "/**\n";
    out << " * @brief  Find an embedded file by path.\n";
    out << " *\n";
    out << " * @param[in]  path     The lookup path (e.g., \"" << (files.empty() ? std::string("Haru/Haru.model3.json") : files[0].relPath) << "\").\n";
    out << " * @param[out] outSize  Receives the file size if found.\n";
    out << " *\n";
    out << " * @return  Pointer to the embedded data, or nullptr if not found.\n";
    out << " */\n";
    out << "static const unsigned char* " << findFunc << "(const char* path, unsigned int* outSize)\n";
    out << "{\n";
    out << "    if (!path) return nullptr;\n\n";
    out << "    // Skip \"live2d/\" prefix if present.\n";
    out << "    const char* searchPath = path;\n";
    out << "    const char* prefix = \"live2d/\";\n";
    out << "    if (strncmp(path, prefix, 7) == 0)\n";
    out << "    {\n";
    out << "        searchPath = path + 7;\n";
    out << "    }\n\n";
    out << "    for (int i = 0; i < " << countName << "; ++i)\n";
    out << "    {\n";
    out << "        if (strcmp(searchPath, " << tableName << "[i].path) == 0)\n";
    out << "        {\n";
    out << "            if (outSize) *outSize = " << tableName << "[i].size;\n";
    out << "            return " << tableName << "[i].data;\n";
    out << "        }\n";
    out << "    }\n";
    out << "    return nullptr;\n";
    out << "}\n\n";

    return out.str();
}

// ------------------------------------------------------------------
// 交互式输入辅助
// ------------------------------------------------------------------

static std::string Trim(const std::string& s)
{
    size_t b = s.find_first_not_of(" \t\r\n");
    if (b == std::string::npos)
        return "";
    size_t e = s.find_last_not_of(" \t\r\n");
    return s.substr(b, e - b + 1);
}

static std::string Prompt(const std::string& question, const std::string& defaultValue)
{
    std::cout << question;
    if (!defaultValue.empty())
        std::cout << " [" << defaultValue << "]";
    std::cout << ": ";
    std::cout.flush();

    std::string line;
    if (!std::getline(std::cin, line))
        return defaultValue;
    line = Trim(line);
    if (line.empty())
        return defaultValue;
    return line;
}

// ------------------------------------------------------------------
// 主程序
// ------------------------------------------------------------------

int main()
{
    std::cout << "============================================================\n";
    std::cout << "  Live2D Model -> C Array Converter (C++ Interactive)\n";
    std::cout << "============================================================\n\n";

    // 1. 模型根目录
    std::string modelRoot;
    while (modelRoot.empty())
    {
        modelRoot = Prompt("模型根目录 (包含模型文件夹的父目录, 如 ./Resources)", "");
#if defined(_WIN32)
        struct _stat st;
        if (_stat(modelRoot.c_str(), &st) != 0 || !(st.st_mode & _S_IFDIR))
#else
        struct stat st;
        if (stat(modelRoot.c_str(), &st) != 0 || !S_ISDIR(st.st_mode))
#endif
        {
            std::cerr << "  [错误] 目录不存在: " << modelRoot << "\n\n";
            modelRoot.clear();
        }
    }

    // 2. 模型名
    std::string modelName = Prompt("模型名 (如 Haru)", "");
    if (modelName.empty())
    {
        std::cerr << "  [错误] 模型名不能为空。\n";
        return 1;
    }

    // 3. 输出头文件路径
    std::string outPath = Prompt("输出头文件路径", "Live2DModelData_" + SanitizeIdentifier(modelName) + ".h");

    // 收集文件
    std::cout << "\n正在扫描目录: " << modelRoot << " ...\n";
    std::vector<FileEntry> files;
    CollectFilesRecursive(modelRoot, modelRoot, outPath, files);
    std::sort(files.begin(), files.end(),
              [](const FileEntry& a, const FileEntry& b) { return a.relPath < b.relPath; });

    if (files.empty())
    {
        std::cerr << "  [错误] 目录下没有找到任何文件: " << modelRoot << "\n";
        return 1;
    }

    std::cout << "找到 " << files.size() << " 个文件:\n";
    for (const auto& fe : files)
        std::cout << "  - " << fe.relPath << "\n";

    // 确认
    std::string confirm = Prompt("\n确认生成头文件?", "y");
    if (confirm != "y" && confirm != "Y" && confirm != "yes" && confirm != "YES")
    {
        std::cout << "已取消。\n";
        return 0;
    }

    // 生成
    std::cout << "\n正在生成头文件...\n";
    std::string header = GenerateHeader(modelName, files);

    std::ofstream f(outPath, std::ios::binary);
    if (!f)
    {
        std::cerr << "  [错误] 无法写入输出文件: " << outPath << "\n";
        return 1;
    }
    f << header;
    f.close();

    // 统计总大小
    unsigned long long total = 0;
    for (const auto& fe : files)
    {
        std::vector<unsigned char> tmp;
        if (ReadFileBytes(fe.absPath, tmp))
            total += tmp.size();
    }

    std::cout << "\n完成!\n";
    std::cout << "  模型名    : " << modelName << "\n";
    std::cout << "  文件数    : " << files.size() << "\n";
    std::cout << "  总大小    : " << total << " 字节 (" << (total / 1024.0 / 1024.0) << " MB)\n";
    std::cout << "  输出      : " << outPath << "\n\n";
    std::cout << "后续步骤:\n";
    std::cout << "  1. 在 LAppPal.cpp (或你的文件加载处) include 这个头文件。\n";
    std::cout << "  2. 在 LAppPal::LoadFileAsBytes() 中调用查找函数, 例如:\n";
    std::cout << "       const unsigned char* d = Find" << SanitizeIdentifier(modelName) << "EmbeddedFile(filePath, &size);\n";
    std::cout << "  3. 用 ndk-build 重新编译。\n";

    return 0;
}

File: tools/convert_model_to_arrays.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
convert_model_to_arrays.py
==========================
把 Live2D 模型目录下的所有文件（.moc3 / .json / .png / .wav / .motion3.json 等）
编译成 C 数组，生成一个 .h 头文件，供 Android NDK 项目从内存直接加载，
无需把文件放进 assets 目录。

用法:
    python3 convert_model_to_arrays.py <模型根目录> <模型名> [输出头文件路径]

示例 (模型根目录 = 包含模型文件夹的父目录):
    python3 convert_model_to_arrays.py ./Resources Haru \
        ../app/src/main/jni/Live2D/Model/Live2DModelData_Haru.h

    ./Resources/
    └── Haru/
        ├── Haru.model3.json
        ├── Haru.moc3
        ├── Haru.2048/texture_00.png
        ├── expressions/F01.exp3.json
        └── motions/haru_g_idle.motion3.json

    生成的查找路径会带 "Haru/" 前缀，例如 "Haru/Haru.model3.json"，
    与 Live2D 运行时使用的路径一致。

生成的头文件包含:
    - 每个文件一个 static const unsigned char 数组
    - 每个文件一个 _size 常量
    - 查找表 Live2DEmbeddedFile g_<model>_embeddedFiles[]
    - 查找函数 Find<Model>EmbeddedFile(path, &size)

注意:
    - 生成的数组是 static 的，请确保头文件只被一个 .cpp 包含（否则每个编译单元
      都会复制一份，导致 .so 体积膨胀）。
    - 若模型文件很大（如 2048 纹理），编译时间会略长，属正常现象。
"""

import os
import re
import sys


def sanitize_identifier(name: str) -> str:
    """把任意路径/文件名转成合法的 C 标识符片段。

    规则: 非字母数字字符 -> 下划线，连续下划线合并，首尾下划线去除。
    例如: "Haru/Haru.2048/texture_00.png" -> "Haru_Haru_2048_texture_00_png"
    """
    s = re.sub(r'[^A-Za-z0-9]+', '_', name)
    s = re.sub(r'_+', '_', s)
    s = s.strip('_')
    return s


def bytes_to_c_array(data: bytes, indent: str = '    ') -> str:
    """把字节串转成 16 字节一行的 C 数组初始化文本。"""
    lines = []
    for i in range(0, len(data), 16):
        chunk = data[i:i + 16]
        hexes = ', '.join('0x%02x' % b for b in chunk)
        lines.append(indent + hexes + ',')
    return '\n'.join(lines)


def collect_files(model_dir: str, exclude: str = None) -> list:
    """递归收集模型目录下所有文件，返回按路径排序的 (相对路径, 绝对路径) 列表。

    exclude: 需要排除的绝对路径（通常是输出头文件本身，防止被递归收集）。
    """
    files = []
    for root, _dirs, names in os.walk(model_dir):
        for name in names:
            full = os.path.join(root, name)
            if exclude and os.path.abspath(full) == os.path.abspath(exclude):
                continue
            rel = os.path.relpath(full, model_dir).replace(os.sep, '/')
            files.append((rel, full))
    files.sort(key=lambda x: x[0])
    return files


def generate_header(model_name: str, files: list) -> str:
    """生成完整的头文件内容。"""
    model_id = sanitize_identifier(model_name)          # 例如 "Haru" -> "Haru"
    model_id_lower = model_id.lower()                   # 例如 "Haru" -> "haru"
    find_func = 'Find%sEmbeddedFile' % model_id         # 例如 "FindHaruEmbeddedFile"
    table_name = 'g_%s_embeddedFiles' % model_id_lower  # 例如 "g_haru_embeddedFiles"
    count_name = 'g_%s_embeddedFileCount' % model_id_lower

    out = []
    out.append('#pragma once')
    out.append('')
    out.append('/**')
    out.append(' * Auto-generated Live2D model data header.')
    out.append(' *')
    out.append(" * All model files for '%s' are embedded as C arrays" % model_name)
    out.append(' * for in-memory loading. No assets directory required at runtime.')
    out.append(' *')
    out.append(' * Do not edit manually -- regenerate with convert_model_to_arrays.py.')
    out.append(' */')
    out.append('')
    out.append('#include <cstring>')
    out.append('#include <cstddef>')
    out.append('')

    entries = []

    for rel, full in files:
        with open(full, 'rb') as f:
            data = f.read()

        # 变量名: g_<model>_<sanitized rel path>，例如 g_haru_Haru_moc3
        # 若相对路径以 "<模型名>/" 开头，则去掉该前缀，避免变量名重复模型名
        # （例如 "Haru/Haru.moc3" -> "Haru.moc3" -> g_haru_Haru_moc3）。
        rel_for_var = rel
        prefix = model_name + '/'
        if rel_for_var.startswith(prefix):
            rel_for_var = rel_for_var[len(prefix):]
        var_name = 'g_%s_%s' % (model_id_lower, sanitize_identifier(rel_for_var))
        size_name = var_name + '_size'

        out.append('// Source: %s' % rel)
        out.append('static const unsigned char %s[] = {' % var_name)
        out.append(bytes_to_c_array(data))
        out.append('};')
        out.append('static const unsigned int %s = %d;' % (size_name, len(data)))
        out.append('')

        entries.append((rel, var_name, size_name))

    # 查找表
    out.append('// ============================================================')
    out.append('// Lookup Table')
    out.append('// ============================================================')
    out.append('')
    out.append('struct Live2DEmbeddedFile {')
    out.append('    const char* path;           // Lookup path (e.g., "%s")' % (entries[0][0] if entries else 'Haru/Haru.model3.json'))
    out.append('    const unsigned char* data;  // Pointer to embedded data')
    out.append('    unsigned int size;           // Size in bytes')
    out.append('};')
    out.append('')
    out.append('static const Live2DEmbeddedFile %s[] = {' % table_name)
    for rel, var_name, size_name in entries:
        out.append('    { "%s", %s, %s },' % (rel, var_name, size_name))
    out.append('};')
    out.append('')
    out.append('static const int %s = %d;' % (count_name, len(entries)))
    out.append('')

    # 查找函数
    out.append('/**')
    out.append(' * @brief  Find an embedded file by path.')
    out.append(' *')
    out.append(' * @param[in]  path     The lookup path (e.g., "%s").' % (entries[0][0] if entries else 'Haru/Haru.model3.json'))
    out.append(' * @param[out] outSize  Receives the file size if found.')
    out.append(' *')
    out.append(' * @return  Pointer to the embedded data, or nullptr if not found.')
    out.append(' */')
    out.append('static const unsigned char* %s(const char* path, unsigned int* outSize)' % find_func)
    out.append('{')
    out.append('    if (!path) return nullptr;')
    out.append('')
    out.append('    // Skip "live2d/" prefix if present.')
    out.append('    const char* searchPath = path;')
    out.append('    const char* prefix = "live2d/";')
    out.append('    if (strncmp(path, prefix, 7) == 0)')
    out.append('    {')
    out.append('        searchPath = path + 7;')
    out.append('    }')
    out.append('')
    out.append('    for (int i = 0; i < %s; ++i)' % count_name)
    out.append('    {')
    out.append('        if (strcmp(searchPath, %s[i].path) == 0)' % table_name)
    out.append('        {')
    out.append('            if (outSize) *outSize = %s[i].size;' % table_name)
    out.append('            return %s[i].data;' % table_name)
    out.append('        }')
    out.append('    }')
    out.append('    return nullptr;')
    out.append('}')
    out.append('')

    return '\n'.join(out)


def main():
    if len(sys.argv) < 3:
        print(__doc__)
        sys.exit(1)

    model_dir = sys.argv[1]
    model_name = sys.argv[2]
    out_path = sys.argv[3] if len(sys.argv) > 3 else 'Live2DModelData_%s.h' % sanitize_identifier(model_name)

    if not os.path.isdir(model_dir):
        print('ERROR: model directory not found: %s' % model_dir)
        sys.exit(1)

    files = collect_files(model_dir, exclude=out_path)
    if not files:
        print('ERROR: no files found in: %s' % model_dir)
        sys.exit(1)

    header = generate_header(model_name, files)

    with open(out_path, 'w', encoding='utf-8') as f:
        f.write(header)

    total = sum(os.path.getsize(full) for _rel, full in files)
    print('Done!')
    print('  Model name : %s' % model_name)
    print('  Files      : %d' % len(files))
    print('  Total size : %d bytes (%.2f MB)' % (total, total / 1024.0 / 1024.0))
    print('  Output     : %s' % out_path)
    print('')
    print('Next steps:')
    print('  1. Include this header in your LAppPal.cpp (or wherever you load files).')
    print('  2. Add a lookup call in LAppPal::LoadFileAsBytes(), e.g.:')
    print('       const unsigned char* d = Find%sEmbeddedFile(filePath, &size);' % sanitize_identifier(model_name))
    print('  3. Rebuild with ndk-build.')


if __name__ == '__main__':
    main()
